## 1. Loading the dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
dfs = pd.read_html(url)
df = dfs[0]
df



Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

### Remove boroughs with 'not assigned' values

In [3]:
df = df[df.Borough != 'Not assigned']
df = df.reset_index(drop=True)
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Let's see how many 'not assigned' Neighbourhoods there are

In [4]:
df[df['Neighbourhood'] == 'not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

#### There are not 'not assigned' values in Neighbourhoods column

In [5]:
print(f'Dataframe has {df.shape[0]} rows and {df.shape[1]} columns')

Dataframe has 103 rows and 3 columns


#### There are currently 103 rows in dataframe

## 2. Getting latitude and longitude coordinates of postal codes

#### I will use the csv file

In [6]:
# The code was removed by Watson Studio for sharing.

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [7]:
df_n = df.merge(df_data_1, how='inner', on='Postal Code')
df_n

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

## 3. Explore and cluster the neighborhoods in Toronto

#### Let's work only with Toronto's Boroughs

In [8]:
toronto_index_list = []
for index, el in enumerate(df_n['Borough'].values):
    if 'toronto' in el.lower():
        toronto_index_list.append(index)
toronto_index_list
df_n = df_n.loc[toronto_index_list, :]
df_n

Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                         Neighbourhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68     Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                   North Toronto West, Lawrence Park  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Moore Park, Summerhill East  43.689574 -79.383160  
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.4000

In [9]:
df_n.shape

(39, 5)

### Segmenting dataframe based on Boroughs so that i can use them to analyze Neighbourhoods

In [11]:
tor_onehot = pd.get_dummies(df_n[['Borough']])
tor_onehot['Neighbourhood'] = df_n['Neighbourhood']
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]
tor_onehot.head()

Neighbourhood  Borough_Central Toronto  \
2                     Regent Park, Harbourfront                        0   
4   Queen's Park, Ontario Provincial Government                        0   
9                      Garden District, Ryerson                        0   
15                               St. James Town                        0   
19                                  The Beaches                        0   

    Borough_Downtown Toronto  Borough_East Toronto  Borough_West Toronto  
2                          1                     0                     0  
4                          1                     0                     0  
9                          1                     0                     0  
15                         1                     0                     0  
19                         0                     1                     0

### Clustering

In [12]:
from sklearn.cluster import KMeans
k_clusters = 4
toronto_clustering = tor_onehot.drop('Neighbourhood', 1)
kmeans = KMeans(n_clusters = k_clusters, random_state=0).fit(toronto_clustering)
labels = kmeans.labels_

#### I will use new df to store the final result

In [13]:
final_df = pd.DataFrame(columns=['Neighbourhood', 'Label', 'Latitude', 'Longitude'])

In [14]:
final_df['Neighbourhood'] = tor_onehot['Neighbourhood']
final_df['Label'] = labels
final_df['Latitude'] = df_n['Latitude']
final_df['Longitude'] = df_n['Longitude']
final_df.reset_index(drop=True, inplace=True)
final_df

Neighbourhood  Label   Latitude  \
0                           Regent Park, Harbourfront      0  43.654260   
1         Queen's Park, Ontario Provincial Government      0  43.662301   
2                            Garden District, Ryerson      0  43.657162   
3                                      St. James Town      0  43.651494   
4                                         The Beaches      3  43.676357   
5                                         Berczy Park      0  43.644771   
6                                  Central Bay Street      0  43.657952   
7                                            Christie      0  43.669542   
8                            Richmond, Adelaide, King      0  43.650571   
9                        Dufferin, Dovercourt Village      2  43.669005   
10  Harbourfront East, Union Station, Toronto Islands      0  43.640816   
11                           Little Portugal, Trinity      2  43.647927   
12                       The Danforth West, Riverdale      3  43.679557   
13           Toronto Dominion Centre, Design Exchange      0  43.647177   
14       Brockton, Parkdale Village, Exhibition Place      2  43.636847   
15                     India Bazaar, The Beaches West      3  43.668999   
16                     Commerce Court, Victoria Hotel      0  43.648198   
17                                    Studio District      3  43.659526   
18                                      Lawrence Park      1  43.728020   
19                                           Roselawn      1  43.711695   
20                                   Davisville North      1  43.712751   
21    Forest Hill North & West, Forest Hill Road Park      1  43.696948   
22                      High Park, The Junction South      2  43.661608   
23                  North Toronto West, Lawrence Park      1  43.715383   
24                The Annex, North Midtown, Yorkville      1  43.672710   
25                             Parkdale, Roncesvalles      2  43.648960   
26                                         Davisville      1  43.704324   
27                     University of Toronto, Harbord      0  43.662696   
28                                 Runnymede, Swansea      2  43.651571   
29                        Moore Park, Summerhill East      1  43.689574   
30          Kensington Market, Chinatown, Grange Park      0  43.653206   
31  Summerhill West, Rathnelly, South Hill, Forest...      1  43.686412   
32  CN Tower, King and Spadina, Railway Lands, Har...      0  43.628947   
33                                           Rosedale      0  43.679563   
34                                     Stn A PO Boxes      0  43.646435   
35                        St. James Town, Cabbagetown      0  43.667967   
36             First Canadian Place, Underground city      0  43.648429   
37                               Church and Wellesley      0  43.665860   
38  Business reply mail Processing Centre, South C...      3  43.662744   

    Longitude  
0  -79.360636  
1  -79.389494  
2  -79.378937  
3  -79.375418  
4  -79.293031  
5  -79.373306  
6  -79.387383  
7  -79.422564  
8  -79.384568  
9  -79.442259  
10 -79.381752  
11 -79.419750  
12 -79.352188  
13 -79.381576  
14 -79.428191  
15 -79.315572  
16 -79.379817  
17 -79.340923  
18 -79.388790  
19 -79.416936  
20 -79.390197  
21 -79.411307  
22 -79.464763  
23 -79.405678  
24 -79.405678  
25 -79.456325  
26 -79.388790  
27 -79.400049  
28 -79.484450  
29 -79.383160  
30 -79.400049  
31 -79.400049  
32 -79.394420  
33 -79.377529  
34 -79.374846  
35 -79.367675  
36 -79.382280  
37 -79.383160  
38 -79.321558

## 4. Visualize 

In [15]:
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors
print('Libraries are ready to use!')

Libraries are ready to use!


#### I will get coordinates of Toronto by web

In [16]:
toronto_latitude = 43.651070
toronto_longitude = -79.347015
map_tor = folium.Map(location=[toronto_latitude, toronto_longitude], zoom_start=10)

In [17]:
x = np.arange(k_clusters)
ys = [i + x + (i*x)**2 for i in range(k_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(final_df['Latitude'], final_df['Longitude'], final_df['Neighbourhood'], final_df['Label']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster],
        fill=True,
        fill_color=rainbow[cluster],
        fill_opacity=0.7).add_to(map_tor)
       
map_tor